WIP commit with the current status of "Analysis with python" project

renaudin opened the pull request:
This is a commit with all the files copied from the repository
used during the project.

There are a jupyter notebook, a script (with almost the same features)
and some plots.

Commit:
First commit with all the files used for the project

This is a commit with all the files copied from the repository
used during the project.

There are a jupyter notebook, a script (with almost the same features)
and some plots.

Commit:
First commit with all the files used for the project

This is a commit with all the files copied from the repository
used during the project.

There are a jupyter notebook, a script (with almost the same features)
and some plots.

renaudin committed:
remove typo

renaudin committed:
remove typo

renaudin committed:
Delete README.md

renaudin committed:
Delete README.md

renaudin committed:
Update Readme.md

renaudin committed:
Update Readme.md

renaudin committed:
add features list for BDT training

 from compare_mc_data.py

renaudin committed:
add features list for BDT training

 from compare_mc_data.py

Commit:
Update of note for AnalysisWithPython

A note describing the goals, process and results of the project is
added.
The notebook is also updated.

Commit:
Update of note for AnalysisWithPython

A note describing the goals, process and results of the project is
added.
The notebook is also updated.

renaudin committed:
Update Readme.md

correct some typos

renaudin committed:
Update Readme.md

correct some typos

renaudin committed:
Update Readme.md

renaudin committed:
Update Readme.md

renaudin committed:
Update Readme.md

renaudin committed:
Update Readme.md

alexpearce commented on contrib-03-AnalysisWithPython/Readme.md:
```diff
@@ -0,0 +1,147 @@
+TMVA.  
+However, it hasn't spread outside of the HEP community while other projects
+also dedicated to data analysis are now very mature with a much larger
+users community.
```
Add a comma “HEP community, while other”.

alexpearce commented on contrib-03-AnalysisWithPython/Readme.md:
```diff
@@ -0,0 +1,147 @@
+However, it hasn't spread outside of the HEP community while other projects
+also dedicated to data analysis are now very mature with a much larger
+users community.
+The Goal of this project was to build an analysis demonstrator capable of :
```
Lowercase “G”.

alexpearce commented on contrib-03-AnalysisWithPython/Readme.md:
```diff
@@ -0,0 +1,147 @@
+
+## Loading data
+
+Using root_pandas library, we can load a ROOT file in a few simple steps :
```
“Using _the_ root_pandas…”

alexpearce commented on contrib-03-AnalysisWithPython/Readme.md:
```diff
@@ -0,0 +1,147 @@
+Using root_pandas library, we can load a ROOT file in a few simple steps :
+- choose the file you want and store the path as string : `myfile = "path/to/file.root"`
+- then select the branches you want by creating a list with the names of the
+  variables you want. The use of of the wildcard **\*** is also supported. For
```
could use backticks here to prevent the need to escape everything, i.e. `*`.

alexpearce commented on contrib-03-AnalysisWithPython/Readme.md:
```diff
@@ -0,0 +1,147 @@
+  number of entries and set, for instance, `n_events = 10000`.
+- Finally, select the name of your tuple as `tuple_name = 'someDir/myTree`
+
+Then, you will be allowed to store you tree as a pandas dataframe using :
```
“_your_ tree”

alexpearce commented on contrib-03-AnalysisWithPython/Readme.md:
````diff
@@ -0,0 +1,147 @@
+- plot an histogram using a selection cut :
+    ```python
+        df.query(cut_string)['B0_M'].plot.hist(bins=custom_binning)
+    ```
````
We could add a caveat here: [`DataFrame.query`](http://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.query.html) will return a copy of the data by default. You can instead generate a ‘view’ by indexing with a boolean mask:

```
mask = (df['B0_M'] > 5250) & (df['B0_P'] > 500)
df_reduced = df[mask]
```

The query string is easier to express, but the `query` call will take up more memory as you end up with two `DataFrame` objects lying around. Which to use depends on the use-case 😄

alexpearce commented on contrib-03-AnalysisWithPython/Readme.md:
````diff
@@ -0,0 +1,147 @@
+```
+## Adding variable to the tree and saving output
+
+Some analysis step could create a new variable, as a BDT output, worth to be
````
“_such_ as a BDT output”, “_which is_ to be added…”

alexpearce commented:
Thank you! This is very neat indeed. I've added a few comments, mostly editorial.

Commit:
update Readme.md taking into account the comments

Commit:
update Readme.md taking into account the comments

renaudin commented:
Thank for your comments, I took them into account and correct a few things. The note is almost ready and Mick and I are still working to give a well documented script.

alexpearce commented:
Thanks for the corrections, LGTM 👍  Looking forward to seeing the script! 👀

apuignav commented on contrib-03-AnalysisWithPython/Readme.md:
```diff
@@ -0,0 +1,153 @@
+## Motivation
+
+In High Energy Physics, almost every data analysis framework involved the ROOT
+software. Built in the 90's, it has been very useful to perform all the tasks
```
Add link to ROOT: https://root.cern.ch/

apuignav commented on contrib-03-AnalysisWithPython/Readme.md:
```diff
@@ -0,0 +1,153 @@
+We tested the following environment on lxplus :
+
+- Install miniconda (or anaconda) for python2.7, Linux 64 bit
+  [link](http://conda.pydata.org/miniconda.html)
```
Maybe give the used commands for the install?

apuignav commented on contrib-03-AnalysisWithPython/Readme.md:
```diff
@@ -0,0 +1,153 @@
+  and the transverse momentum of each particle of your entry.
+- As the file will be loaded in the memory, you should limit to a certain
+  number of entries and set, for instance, `n_events = 10000`.
+- Finally, select the name of your tuple as `tuple_name = 'someDir/myTree`
```
Maybe the code corresponding to all these steps could be moved to its own code block after the enumeration with its own comments to help:

``` python
# Path of the ROOT file
myfile = "path/to/file.root"
# Branches to load, wildcards allowed
mybranches = ["B0_M", "*_PT"]
# Number of entries to load
n_events = 10000
# Name of the tree
tuple_name = "someDir/myTree"
```

apuignav commented on contrib-03-AnalysisWithPython/Readme.md:
````diff
@@ -0,0 +1,153 @@
+Then, you will be allowed to store your tree as a pandas dataframe using :
+```python
+   import root_pandas
+   df = root_pandas.read_root(myfile, columns=mybranches, stop = n_events, key= tuple_name)
````
PEP8: `df = root_pandas.read_root(myfile, columns=mybranches, stop=n_events, key=tuple_name)`

apuignav commented on contrib-03-AnalysisWithPython/Readme.md:
````diff
@@ -0,0 +1,153 @@
+```python
+   import numpy as np
+   custom_binning = np.linspace(5000,5500,100) #min, max and number of bins 
+   mass_B0 = df['B0_M'].plot.hist(bins=custom_binning, label = 'mass of the
````
I think it's better if you use `B0_MM`, that's the recommended mass to use.

alexpearce replied:
No, it's almost always the `_M` branch one should use. There was a [thread on `lhcb-tesla` about it](https://groups.cern.ch/group/lhcb-tesla/Lists/Archive/Flat.aspx?RootFolder=%2Fgroup%2Flhcb-tesla%2FLists%2FArchive%2Fdifference%20between%20XX_MM%20and%20XX_M&FolderCTID=0x01200200D58CE2C7A01AE94382C3D6890D64B854) a while ago, where Vanya said:

> Honestly speaking I personally do not understand why people are using MM at all...  - it could be useful for unndertsanding of vertex constrains, but fro all physics purposed, clearly  only "M" should be used  (or , much better, mass  after full DTF-refit)

apuignav commented on contrib-03-AnalysisWithPython/Readme.md:
````diff
@@ -0,0 +1,153 @@
+```
+## Creating and plotting histograms
+
+To produce a simple plot, let's say, the mass of the B0, you can simply do :
````
Use LaTeX for the B0 -> $B^{0}$

apuignav commented on contrib-03-AnalysisWithPython/Readme.md:
```diff
@@ -0,0 +1,153 @@
+`AnalysisWithPython.ipynb`.  
+
+
+## Cutting
```
Cutting -> Selecting events

In principle, "cut" is jargon so we should try to avoid it as much as possible

apuignav commented on contrib-03-AnalysisWithPython/Readme.md:
````diff
@@ -0,0 +1,153 @@
+    mask = (df['B0_M'] > 5250) & (df['B0_P'] > 500)
+    df_reduced = df[mask]
+    ```
+Which method to choose will depend on the use-case. And then, the plotting part
````
Can you expand in the pros and cons of each method?

apuignav commented on contrib-03-AnalysisWithPython/Readme.md:
```diff
@@ -0,0 +1,153 @@
+    - `source activate NameOfYourVirtualEnv` if you made it yourself
+- A few packages are also needed and can be installed using pip, the python
+  package manager :
+    - Numpy (arrays and tools for scientific computing), Matplotlib (plotting
```
References (links) to all packages are needed

apuignav commented on contrib-03-AnalysisWithPython/Readme.md:
```diff
@@ -0,0 +1,153 @@
+added to a ROOT file. Two steps are needed to perform this :
+- The variable should be added to the dataframe using `df['BDT'] = sk_y_predicted`
+- The modified dataframe can be stored in a new ROOT file using
+  `df.root_pandas.to_root(df, 'output.root', key = 'mytree')`
```
Same comment as before: maybe add the code as a block at the end of the enumeration?

``` python
df['BDT'] = sk_y_predicted
df.root_pandas.to_root(df, 'output.root', key='mytree')
```

apuignav replied:
Could you add a few lines on how to store the extra variable back to disk, either as a ROOT file or as a Pandas dataframe? That would close the loop.

apuignav commented on contrib-03-AnalysisWithPython/Readme.md:
```diff
@@ -0,0 +1,153 @@
+- The modified dataframe can be stored in a new ROOT file using
+  `df.root_pandas.to_root(df, 'output.root', key = 'mytree')`
+
+## Conclusion
```
Before the conclusion, I think it would be nice to have a "Going further" (short) section in which you outline how to continue with the analysis, maybe giving links to useful tools for python-based analysis.

apuignav commented:
Great work! I added my comments inline.

As a final comment, maybe we could change the title to "Getting started with python-based analysis"? This is more adequate to the content and makes it maybe more useful for people reading the proceedings.

alexpearce commented:
:+1: on the title change.

renaudin commented:
I will look at the thread you mentionned, and I agree, the proposed title is much better.

renaudin commented:
About the Going further section and adding links to good material, I was thinking about pointing to :
- Software Carpentry lessons
- Starterkit lessons ;)
- Tim's blog

Do you have other links you would like to see in this section ?

apuignav commented:
I'm not sure these are the most useful links. For full python-based analysis, maybe
- Theano/tensorflow
- Matplotlib
- The Impactkit Papermake article (#1)

What else? @ibab, @alexpearce?

alexpearce commented:
I would add:
- [scipy](https://www.scipy.org/) for general scientific computation and numerical routines (such as kernel density estimate, spline interpolation, minimisation, etc.)
- The [uncertainties](https://pythonhosted.org/uncertainties/) and/or [mcerp](https://github.com/tisimst/mcerp) packages for dealing with error propagation
- The [iminuit](https://github.com/iminuit/iminuit) and/or [probfit](https://github.com/iminuit/probfit) packages for fitting

I think Theano/Tensorflow might be a bit daunting to start with.

apuignav commented:
Completely agree with the list. Maybe we could add [soerp](https://pypi.python.org/pypi/soerp) for error propagation too?

alexpearce commented:
I haven't used soerp, but we could add it (I think mcerp trumps just about anything, but it might be good to show the choice…).

apuignav commented:
Never used either, I'm an `uncertainties` guy (I even wrote some code for it), but now I will have to try 😄

Commit:
Updated version of the note

The comments of Alex and Albert were taken into account

A better commented version of the python script is also shipped

Commit:
Updated version of the note

The comments of Alex and Albert were taken into account

A better commented version of the python script is also shipped

Commit:
Update of script and notebook

pep8 compliance
more coherence between notebook and script

Commit:
Update of script and notebook

pep8 compliance
more coherence between notebook and script

Commit:
commit for HOP LoKi Functor (#4)

TupleTool and LoKi functors for HOP mass

Commit:
Merge branch 'master' of github.com:renaudin/impactkit-proceedings
